File: S2_07052025/ConvertPython/Ex2_Python/Ex2_2.py
import cv2
import matplotlib.pyplot as plt
import numpy as np

img = cv2.imread('tiger.jpg', cv2.IMREAD_GRAYSCALE).astype(np.float64)
plt.figure(figsize=(15, 6))
for i in range(1, 9):
    levels = 2 ** i
    step = 256 / levels
    quantized_img = np.ceil(img / step) * step - 1
    quantized_img = np.clip(quantized_img, 0, 255).astype(np.uint8)
    plt.subplot(2, 4, i)
    plt.imshow(quantized_img, cmap='gray')
    label = f"{i}-bit" if i == 1 else f"{i}-bits"
    plt.title(label, fontsize=15)
    plt.axis('off')
    cv2.imwrite(f"Quantization_in_{label}.jpg", quantized_img)
plt.show()

File: self/S2_07052025/Ex2_2.m
% Digitalzation: Sampling and Qiantization
clear, clc, close all;
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
pkg load image; %load image package in Octave

%Load test image
Img = double(imread('tiger.jpg'));

%díplay the original image with different number of bits
FS = 15;
for NumOfBit = 1 : 8
  %quantize image to a given number of bit(s)
  NumOfLevel = 2 .^ NumOfBit; %number of levels
  LevelGap = 256 / NumOfLevel; %different between two adjacent levels
  quantized_Img = uint8(ceil(Img / LevelGap) * LevelGap - 1);

  %display image
  subplot(2, 4, NumOfBit);
  imshow(quantized_Img);

  %image caption
  if (1 == NumOfBit)
    name = [num2str(NumOfBit) '-bit'];
  else
    name = [num2str(NumOfBit) '-bits'];
  endif
  title(name, "fontsize", FS);

  %save image
  imwrite(quantized_Img, ['Quantization_in_' name '.jpg']);
end
